lib/saml/url_service.rb

```diff
@@ -17,27 +17,44 @@ class URLService
     LOGIN_REDIRECT_PARTIAL = '/auth/login/callback'
     LOGOUT_REDIRECT_PARTIAL = '/logout/'
 
-    attr_reader :saml_settings, :session, :authn_context
+    attr_reader :saml_settings, :session, :user, :authn_context, :type, :client_id
 
-    def initialize(saml_settings, session: nil, user: nil)
+    def initialize(saml_settings, session: nil, user: nil, params: {})
       if session.present?
         @session = session
+        @user = user
         @authn_context = user&.authn_context
       end
+
       @saml_settings = saml_settings
+
+      if params[:action] == 'new'
+        @type = params[:type]
+        @client_id = params[:client_id]
+      elsif params[:action] == 'saml_callback'
+        @type = JSON.parse(params[:RelayState])[:type] if params[:RelayState]
+      end
     end
 
     # REDIRECT_URLS
     def base_redirect_url
       VIRTUAL_HOST_MAPPINGS[current_host][:base_redirect]
     end
 
-    def login_redirect_url(params = {})
-      add_query("#{base_redirect_url}#{LOGIN_REDIRECT_PARTIAL}", params)
+    def login_redirect_url(auth: 'success', code: nil)
+      if auth == 'success' && user.loa[:current] < user.loa[:highest]
+        verify_url
+      else
+        query_params = {}
+        query_params[:type] = type if type
+        query_params[:auth] = auth if auth == 'fail'
+        query_params[:code] = code if code
+        add_query("#{base_redirect_url}#{LOGIN_REDIRECT_PARTIAL}", query_params)
+      end
     end
 
-    def logout_redirect_url(params = {})
-      add_query("#{base_redirect_url}#{LOGOUT_REDIRECT_PARTIAL}", params)
+    def logout_redirect_url
+      "#{base_redirect_url}#{LOGOUT_REDIRECT_PARTIAL}"
     end
```

Comment:
If we want to always directly pass params from `SessionController` to `URLService`, things like `auth` and `code` will not come through, because they're not part of the params received in `SessionController`. We'd have to merge them into the params being passed to the service at initialization.